core/src/main/java/io/substrait/relation/ExpressionCopyOnWriteVisitor.java

```diff
@@ -0,0 +1,365 @@
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.BinaryLiteral expr) throws EXCEPTION {
+    return visitLiteral(expr);
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.TimeLiteral expr) throws EXCEPTION {
+    return visitLiteral(expr);
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.DateLiteral expr) throws EXCEPTION {
+    return visitLiteral(expr);
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.TimestampLiteral expr) throws EXCEPTION {
+    return visitLiteral(expr);
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.TimestampTZLiteral expr) throws EXCEPTION {
+    return visitLiteral(expr);
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.IntervalYearLiteral expr) throws EXCEPTION {
+    return visitLiteral(expr);
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.IntervalDayLiteral expr) throws EXCEPTION {
+    return visitLiteral(expr);
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.UUIDLiteral expr) throws EXCEPTION {
+    return visitLiteral(expr);
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.FixedCharLiteral expr) throws EXCEPTION {
+    return visitLiteral(expr);
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.VarCharLiteral expr) throws EXCEPTION {
+    return visitLiteral(expr);
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.FixedBinaryLiteral expr) throws EXCEPTION {
+    return visitLiteral(expr);
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.DecimalLiteral expr) throws EXCEPTION {
+    return visitLiteral(expr);
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.MapLiteral expr) throws EXCEPTION {
+    return visitLiteral(expr);
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.ListLiteral expr) throws EXCEPTION {
+    return visitLiteral(expr);
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.StructLiteral expr) throws EXCEPTION {
+    return visitLiteral(expr);
+  }
+
+  @Override
+  public Optional<Expression> visit(Expression.Switch expr) throws EXCEPTION {
+    var match = expr.match().accept(this);
+    var switchClauses = transformList(expr.switchClauses(), this::visitSwitchClause);
+    var defaultClause = expr.defaultClause().accept(this);
+
+    if (allEmpty(match, switchClauses, defaultClause)) {
+      return Optional.empty();
+    }
+    return Optional.of(
+        Expression.Switch.builder()
+            .from(expr)
+            .match(match.orElse(expr.match()))
+            .switchClauses(switchClauses.orElse(expr.switchClauses()))
+            .defaultClause(defaultClause.orElse(expr.defaultClause()))
+            .build());
+  }
+
+  protected Optional<Expression.SwitchClause> visitSwitchClause(
+      Expression.SwitchClause switchClause) throws EXCEPTION {
+    return switchClause
+        .then()
```

Comment:
Does this not need to also visit over the `switchClause.condition()`?

Reply:
I thought about this and wasn't sure of what the correct behaviour should be actually.

The condition clause is a `Literal` and not an `Expression`, so it's guaranteed not to contain an `Expression`.

While `Literal` does extend `Expression`, meaning we can call `.accept` on it, the visitor can return _any_ kind of `Expression` and not just a `Literal` which feels like it could break in subtle ways.

Part of my intent with having an explicit `visitSwitchClause` method was to allow users to override it if they wanted to muck with the condition themselves.

Reply:
> the visitor can return any kind of Expression and not just a Literal which feels like it could break in subtle ways.

ah!  This is a great point.  I wonder if it would be worth a comment to that effect for future readers.